arbor/profile/profiler.cpp

```diff
@@ -169,17 +178,33 @@ profiler::profiler() {
     recorders_.resize(threading::num_threads());
 }
 
+profiler::profiler(unsigned nthreads) {
+    recorders_.resize(nthreads);
+}
+
+void profiler::initialize(task_system_handle& ts) {
+    recorders_.resize(ts.get()->get_num_threads());
+    thread_ids_ = ts.get()->get_thread_ids();
+    init_ = true;
+}
+
 void profiler::enter(region_id_type index) {
-    recorders_[threading::thread_id()].enter(index);
+    if(!init_)
+        throw arbor_exception(util::pprintf(
+                "Profiler not initialized with correct task_system parameters"));
+    recorders_[thread_ids_.at(std::this_thread::get_id())].enter(index);
 }
 
 void profiler::enter(const char* name) {
+    if(!init_)
```

Comment:
How about just returning if uninitialized? That way if the application has been compiled with profiler support on, but the user does not initialize the profiler, nothing is recorded.